src/software/ai/evaluation/intercept.h

```diff
@@ -19,5 +19,25 @@
  *         relative to the timestamp of the robot. If no possible intercept could be
  * found within the field bounds, returns std::nullopt
  */
-std::optional<std::pair<Point, Duration>> findBestInterceptForBall(Ball ball, Field field,
-                                                                   Robot robot);
+std::optional<std::pair<Point, Duration>> findBestInterceptForBall(
+    TimestampedBallState ball, Field field, TimestampedRobotState robot);
+
+/**
+ * Returns the estimated position of the ball at a future time, relative to when the
+ * ball was last updated.
+ *
+ * @param position current position of the ball
+ * @param velocity current velocity of the ball
+ * @param duration_in_future The relative amount of time in the future
+ * at which to predict the ball's position. Value must be >= 0.
+ * For example, a value of 1.5 seconds would return the estimated position of the ball
+ * 1.5 seconds in the future.
+ *
+ * @throws std::invalid_argument if the ball is estimating the position with a time
+ * from the past
+ *
+ * @return the estimated position of the ball at the given number of milliseconds
+ * in the future. Coordinates are in metres.
+ */
+Point estimateBallPositionAtFutureTime(Point position, Vector velocity,
+                                       const Duration &duration_in_future);
```

Comment:
Looking at this again, I'm very confused why this function is not a part of the ball? It seems pretty logical that any ball model would be a part of the ball class?

Reply:
IMO it would seem weird that the Ball is responsible for predicting its own future state.

As soon as we get a model that is more accurate than speed*time - like fiction etc, I don't think it makes sense for the Ball class to have knowledge of that kind of information. @MathewMacDougall

Reply:
I agree with Evan's that estimate future time is a utility function (think geom types) that doesn't belong in `Ball`. The longterm goal of this rearch is that Ball will become effectively TimestampedBallState and the current functionality will live in evaluation functions or in SensorFusion filters. This PR is already 1500 lines and I haven't changed anything functional. I thought it would be a good place to stop before I move more things around. With this PR a lot of things in Ball and Robot are dismantled. `Ball::esimatePositionAtFutureTime` was only being used in intercept so I pulled it out here. I can inline it if you feel that's an improvement.

Reply:
Hmmm I see the arguments for both sides here (function in/outside of the class). Trying to separate my bias of being "familiar" with it as part of the class.

If for example we stored a ball model in some Dynamic Params in the future, those could easily be passed to the Ball class or a free function, so I don't see a huge advantage one way or the either there.

Would be good to hear from @LiCody @jared-paul or @kelvinkoon as to what they find more intuitive too.

Reply:
I think what we agreed on was that `World` would contain a lightweight data type to represent the ball, e.g. `TimestampedBallState` and it would not contain history or any additional functionality (same for TimestampedRobotState). Ideas like possession and BallModel should be handled separately. So BallModel can be its own class if that makes sense when we implement it, but it won't be the ball class that gets passed from SensorFusion to AI.

Reply:
I might have misspoke during our discussion, but my intent was that `World` holds an abstraction around the ball, including being able to get it's estimated future state. I was advocating for the removal of history, which seemed like an implementation detail.

Reply:
I feel like adding state prediction to World would require a name change to the class. 'World' implies a world state at a moment vs fluidity of time IMO.

Reply:
I mean, we need to somehow convey this model to the AI without the raw buffers being exposed.

Reply:
Why does ball prediction need inputs from a state history buffer? I thought that ball prediction can be done statically from a physics model of the ball and the ball state (velocity + position).

Reply:
There are some ball prediction algorithms that would require history. Based solely on friction we could determine future states from the current state, but any curvature (due to rotating dribblers/kickers) you would need previous states to judge the curve

Reply:
That's a good point since we'll have rotating kickers soon. I think @MathewMacDougall point that we want BallState to be as simple as possible for so we don't have to mock things that we don't need is relevant here. SensorFusion should define a separate BallModel class to pass into AI rather than piggyback on ball state. I can work on that structure in #1522

Reply:
Bit confused here, are we suggesting that `World` would have `BallState` and `BallModel` as independent objects? 

(_might be faster to have this discussion on slack_)

Reply:
An important point here is to keep `World` to only representing a snapshot of the world, since it is used in `Simulation`. After offline discussion, there will be a need to combine `world_state`, `possession`, `ball_model`, and other modules into a `SensorFusion` output for `AI` to use. `estimateBallPositionAtFutureTime` will be moved once when we work on the `BallModel`

Reply:
I can't comment on whether or not this is needed for Sim (though I suspect it's not), but I'm really not sure where this is going in terms of the output of sensor fusion. I certainly don't think it makes any sense to break up the `Ball` class, perhaps it might make sense to have `Ball` contain a `BallState` and a `BallModel`?

Reply:
> An important point here is to keep World to only representing a snapshot of the world, since it is used in Simulation

World has no bearing on simulation. Simulation just outputs raw SSL data, it has no knowledge of the World.

I think `Ball` containing a ballstate and ball model makes sense, because it puts it at a similar level of abstraction to the new (soon TM) Robot class that is basically state with "extra metadata"

Reply:
OK that makes sense, I'll make a BallModel and put it in Ball in the next PR with #1522